lib/Differentiator/ReverseModeVisitor.cpp

```diff
@@ -295,7 +346,119 @@ namespace clad {
+
+      beginScope(Scope::FunctionPrototypeScope |
+                 Scope::FunctionDeclarationScope | Scope::DeclScope);
+      m_Sema.PushFunctionScope();
+      m_Sema.PushDeclContext(getCurrentScope(), gradientOverloadFD);
+
+      llvm::SmallVector<ParmVarDecl*, 4> overloadParams;
+      llvm::SmallVector<Expr*, 4> callArgs;
+
+      overloadParams.reserve(FD->getNumParams() * 2);
+      callArgs.reserve(paramTypes.size());
+
+      for (auto* GVD : params) {
+        auto* VD = ParmVarDecl::Create(
+            m_Context,
+            gradientOverloadFD,
+            noLoc,
+            noLoc,
+            GVD->getIdentifier(),
+            GVD->getType(),
+            GVD->getTypeSourceInfo(),
+            GVD->getStorageClass(),
+            // Clone default arg if present.
+            (GVD->hasDefaultArg() ? Clone(GVD->getDefaultArg()) : nullptr));
+        if (VD->getIdentifier())
+          m_Sema.PushOnScopeChains(VD,
+                                   getCurrentScope(),
+                                   /*AddToContext*/ false);
+        overloadParams.emplace_back(VD);
+        callArgs.emplace_back((Expr*)BuildDeclRef(VD));
+      }
+
+      QualType DVDType;
+      DVDType = m_Context.getPointerType(m_Function->getReturnType());
+      for (int i = 0; i < remainingArgs; i++) {
+        IdentifierInfo* DVDII =
+            &m_Context.Idents.get("_d_" + std::to_string(i));
+
+        auto DVD = ParmVarDecl::Create(
+            m_Context,
+            gradientOverloadFD,
+            noLoc,
+            noLoc,
+            DVDII,
+            DVDType,
+            m_Context.getTrivialTypeSourceInfo(DVDType, noLoc),
+            StorageClass::SC_None,
+            // Clone default arg if present.
+            nullptr);
+        if (DVD->getIdentifier())
+          m_Sema.PushOnScopeChains(DVD,
+                                   getCurrentScope(),
+                                   /*AddToContext*/ false);
+        overloadParams.emplace_back(DVD);
+      }
+
+      llvm::ArrayRef<ParmVarDecl*> overloadParamsRef =
+          llvm::makeArrayRef(overloadParams.data(), overloadParams.size());
+
+      llvm::MutableArrayRef<Expr*> callArgsRef =
+          llvm::makeMutableArrayRef(callArgs.data(), callArgs.size());
+
+      gradientOverloadFD->setParams(overloadParamsRef);
+      gradientOverloadFD->setBody(nullptr);
+
+      beginScope(Scope::FnScope | Scope::DeclScope);
+      m_DerivativeFnScope = getCurrentScope();
+      beginBlock();
+
+      Expr* callExpr = nullptr;
+      if (auto* gradientCMD = dyn_cast<CXXMethodDecl>(gradientFD)) {
+        callExpr = BuildCallExprToMemFn(gradientCMD, callArgsRef);
+      } else {
+        assert(isa<FunctionDecl>(gradientFD) && "Unexpected!");
+        callExpr = BuildCallExprToFunction(gradientFD, callArgsRef);
+      }
+      addToCurrentBlock(callExpr);
+      Stmt* gradientOverloadBody = endBlock();
+
+      gradientOverloadFD->setBody(gradientOverloadBody);
+    }
```

Comment:
I believe we are missing ending scopes here. 
```suggestion
      endScope(); // Function body scope
      m_Sema.PopFunctionScopeInfo();
      m_Sema.PopDeclContext();
      endScope(); // Function decl scope
    }
```